File: client/src/main/java/pro/panopticon/client/sensor/impl/SingleCallSensor.java
package pro.panopticon.client.sensor.impl;

import pro.panopticon.client.model.Measurement;
import pro.panopticon.client.sensor.Sensor;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class SingleCallSensor implements Sensor {

    private final Map<AlertInfo, Status> measurements;

    public SingleCallSensor() {
        this.measurements = new HashMap<>();
    }

    @Override
    public List<Measurement> measure() {
        return measurements.entrySet().stream().map((Map.Entry<AlertInfo, Status> it) ->
                new Measurement(
                        it.getKey().getSensorKey(),
                        getPanopticonStatus(it.getValue()),
                        getDisplayValue(it.getValue()),
                        it.getKey().getDescription()
                )
        ).collect(Collectors.toList());

    }

    public void triggerError(AlertInfo alertInfo) {
        measurements.compute(alertInfo, (k, v) -> Status.ERROR);
    }

    public void triggerWarn(AlertInfo alertInfo) {
        measurements.compute(alertInfo, (k, v) -> Status.WARN);
    }
    
    public void triggerOk(AlertInfo alertInfo) {
        measurements.compute(alertInfo, (k, v) -> Status.OK);
    }

    private String getPanopticonStatus(Status status) {
        switch (status) {
            case ERROR:
                return "ERROR";
            case WARN:
                return "WARN";
            case OK:
                return "INFO";
            default:
                return "";
        }
    }

    private String getDisplayValue(Status status) {
        switch(status) {
            case ERROR:
                return "In error";
            case WARN:
                return "Status: WARN";
            case OK:
            default:
                return "Status: OK";
        }
    }

    private enum Status {
        OK, WARN, ERROR
    }
}

File: client/src/main/java/pro/panopticon/client/sensor/impl/SingleCallSensor.kt
package pro.panopticon.client.sensor.impl

import pro.panopticon.client.model.Measurement
import pro.panopticon.client.model.Measurement.Status
import pro.panopticon.client.sensor.Sensor
import java.util.HashMap

open class SingleCallSensor : Sensor {
    private val measurements: MutableMap<Sensor.AlertInfo, Status>
    override fun measure(): List<Measurement> {
        return measurements.entries.map {
            Measurement(
                key = it.key.sensorKey,
                status = it.value,
                displayValue = getDisplayValue(it.value),
                description = it.key.description,
            )
        }
    }

    open fun triggerError(alertInfo: Sensor.AlertInfo) {
        measurements.compute(alertInfo) { _: Sensor.AlertInfo?, _: Status? -> Status.ERROR }
    }

    open fun triggerWarn(alertInfo: Sensor.AlertInfo) {
        measurements.compute(alertInfo) { _: Sensor.AlertInfo?, _: Status? -> Status.WARN }
    }

    open fun triggerOk(alertInfo: Sensor.AlertInfo) {
        measurements.compute(alertInfo) { _: Sensor.AlertInfo?, _: Status? -> Status.INFO }
    }

    private fun getDisplayValue(status: Status): String {
        return when (status) {
            Status.ERROR -> "In error"
            Status.WARN -> "Status: WARN"
            Status.INFO -> "Status: OK"
        }
    }

    init {
        measurements = HashMap()
    }
}
